Sets current narrow marker for empty narrows.

sumanthvrao opened the pull request:
This PR address setting labels in narrow's with no previous message. A dummy message is created for such narrows with Welcome-Bot as the author and a suitable content, This message gets discarded when the first message arrives. Further, it contains tests for the added features.
- [x] In streams with no previous messages. 
![new_stream_message](https://user-images.githubusercontent.com/30312566/52068277-602cf700-25a2-11e9-8ece-38f83eb98138.png)

- [x] In PM with no previous messages.
![post2](https://user-images.githubusercontent.com/30312566/52027980-4a341d80-2533-11e9-9254-2d980bccb2de.png)

Fixes : #259, #266

neiljp reviewed (commented):
@sumanthvrao I've left some preliminary thoughts, though this is mostly on coding style rather than the approach to fixing the problem - though I do see it works in practice, based on manual testing.

I know you mentioned this doesn't work for muted streams, but it also doesn't for other types of general empty narrows, including eg. PMs (select a user in the right-hand list, for example).

I'm wondering if these changes can be split into smaller parts (and commits) which can then be connected in a later commit?

You might find adding some tests would make it clearer to yourself (and to reviewers) what features you're adding, as you can eg. add a parameter and a test for that function with the parameter, often alone in one commit.

I'm thinking we should show a 'special message' (ie. message 'content') rather than an empty message with a header, if we go this route; what do you think? (also @amanagr and anyone else)

sumanthvrao commented on zulipterminal/core.py:
```diff
@@ -132,9 +132,11 @@ def narrow_to_stream(self, button: Any) -> None:
         else:
-            w_list = create_msg_box_list(self.model, msg_id_list)
+            w_list = create_msg_box_list(self.model, msg_id_list,
+                                         stream_cap=button.caption)
```
Yup. Done

sumanthvrao commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -120,13 +120,15 @@ def keypress(self, size: Tuple[int, int], key: str) -> str:
         self.title = ''
         self.email = ''
+        # if this is the only message in the stream
+        self.empty_stream = empty_stream
```
True. i renamed it to `is_in_empty_stream`. Did away with the comment. As you said this makes better reading now.

sumanthvrao commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -149,6 +151,8 @@ def _time_for_message(self, message: Dict[str, Any]) -> str:
         return ctime(message['timestamp'])[:-8]
 
     def need_recipient_header(self) -> bool:
+        if(self.empty_stream):
```
This is kind off a habit for me :')

neiljp replied:
Coming from a language like C/C++? :)

sumanthvrao replied:
Yup! :)

sumanthvrao commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -334,6 +338,11 @@ def main_view(self) -> List[Any]:
+        # Dummy messages with just headers and no content
+        if self.empty_stream:
+            parts = [(recipient_header, recipient_header is not None)]
+            return [part for part, condition in parts if condition]
```
Since i initially though we could create just a dummy message, which has only a recipient header, we don't need content, and other properties. Returning it here itself helps us avoid filling different properties of the dummy message like `author`, `timestamp`.
I did change it to `return [receipient header]` which does the same thing that code does.

neiljp replied:
I'd need to read it again, and I'm off to bed, but I'm not sure that code is equivalent - if `recipient_header is not None` then we get `[recipient_header]` as you suggest, but when the condition is reversed I'm not sure they are equivalent.

sumanthvrao replied:
Yes. But i doubt the condition will ever get reversed if execution reaches till there. Anyway i'll check it again.

sumanthvrao commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -461,19 +470,19 @@ def keypress(self, size: Tuple[int, int], key: str) -> str:
         elif is_command_key('GO_BACK', key):
             self.model.controller.show_all_messages(self)
-        elif is_command_key('REPLY_AUTHOR', key):
+        elif is_command_key('REPLY_AUTHOR', key) and not self.empty_stream:
```
Yup, I will write tests on this after which it will make more sense.

sumanthvrao commented on zulipterminal/ui_tools/utils.py:
```diff
@@ -18,11 +19,27 @@ def create_msg_box_list(model: Any, messages: Union[None, Iterable[Any]]=None,
+        msg['subject'] = ''
+        msg['id'] = None  # type: ignore
+        msg['reactions'] = []  # type: ignore
+        msg['flags'] = ['read']  # type: ignore
```
Thanks. Yes. I did make it inline.

sumanthvrao commented on .gitignore:
```diff
@@ -4,6 +4,7 @@ zuliprc
 cov_html/*
 .coverage
 .mypy_cache
+.mypy-cache
```
The cache file created was `.mypy-cache` and not `mypy_cache` due to which i added it to gitignore.

sumanthvrao replied:
True. I'll have a look.

sumanthvrao commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -170,6 +173,15 @@ def need_recipient_header(self) -> bool:
         else:
             raise RuntimeError("Invalid message type")
 
+    def _check_if_self_PM(self) -> bool:
```
This is for the bug fix when you sent a PM to yourself. If you wanted to reply to a message( mouse clicking on it / `r`/ ENTER), a fotter is created. The email address in 'To:` would be blank for such PM's with oneself.

neiljp reviewed (commented):
This is quite a large PR and most is in one commit, which makes it difficult to understand in one go. I've left some comments, but with some refactoring and addition of the features in multiple commits it should be easier to understand.

neiljp commented on zulipterminal/ui_tools/utils.py:
```diff
@@ -18,8 +21,47 @@ def create_msg_box_list(model: Any, messages: Union[None, Iterable[Any]]=None,
+    # no previous messages, whose author is welcome bot.
+    if (message_list == [] and (stream_caption is not None or
+                                pm_details is not None)):
+        msg = {}
```
To clarify, rather than
```python3
x = {}
if y:
    x.update(dict1)
elif z:
    x.update(dict2)
```
it may be clearer to have:
```python3
if y:
    x = dict1
elif z:
    x = dict2
```

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -117,13 +117,14 @@ def keypress(self, size: Tuple[int, int], key: str) -> str:
 
 class MessageBox(urwid.Pile):
```
You don't need to name the commit summary as 'unfactored'; only describe what each does. The refactoring commit can start with 'refactor <something>...', but you're describing a change in each case, rather than a 'state'.

I was expecting to see tests in this commit, to go with the extra parameter.

neiljp commented on tests/ui/test_ui_tools.py:
```diff
@@ -1070,6 +1078,58 @@ def test_main_view_generates_PM_header(self, mocker, message,
+        }, True)
+    ])
+    def test_dummy_message(self, mocker, message,
+                           is_empty_narrow):
```
As I said re the other commit, I would expect this test to go in the commit with the feature introduction, not the code refactoring. Generally when refactoring, tests should pass except minor changes - when you add a new feature/fix, then you add more tests to lock that new aspect in place.

neiljp commented on tests/model/test_model.py:
```diff
@@ -513,6 +513,38 @@ def test_append_message_with_valid_log(self, mocker, model):
 
+    @pytest.mark.parametrize('message', [
+        ({
+            'type': 'private',
```
The 'private' and 'stream' are the only changes here, and the rest is duplicated; I would suggest constructing the message in the test and parametrizing based on the type only.

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -167,6 +170,15 @@ def need_recipient_header(self) -> bool:
+                self.model.user_email:
+            return True
+        else:
+            return False
```
Whenever one returns a boolean, it's worth considering whether to return a value explicitly (as you do above) or returning the value of an expression; can you see how you could do the latter here? Would it be simpler?

neiljp reviewed (commented):
Thanks for doing the work on splitting this - it's a lot more readable post-splitting :+1: If you can place the refactoring before the features/fixes, and keep the test changes with each code change, that will make this even more easily reviewable - ie. make the 'story' clearer :)

Note that you can have multiple lines in the text after the commit title, so you can eg. discuss the what/why you changed in one paragraph, and then have a blank line before another sentence, if those things are not related (like that you included test changes).

neiljp commented on zulipterminal/model.py:
```diff
@@ -492,6 +492,28 @@ def append_message(self, event: Event) -> None:
+            if self.msg_list.log:
+                possible_dummy_message = \
+                    self.msg_list.log[0].original_widget.message
+                if possible_dummy_message.get('id') is None:
```
Did you choose `.get` here intentionally, over indexing? We're checking for the special value for when it's the dummy case, right?

neiljp commented on zulipterminal/model.py:
```diff
@@ -520,6 +542,9 @@ def append_message(self, event: Event) -> None:
                 if recipients == msg_recipients:
                     self.msg_list.log.append(msg_w)
 
+            if(not dummy_message_found):
```
Other than the unnecessary parentheses, what is this for?

sumanthvrao replied:
So this is as per conversation https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/PR.20.23T278/near/696566. We don't want new messages popping up in same narrow. ( This doesn't delete the message, only avoids displaying it)

neiljp reviewed (commented):
Good work with the tests, and thanks for this iteration; I hope the review comments are useful to you :)

neiljp commented:
NOTE: The 'enable dummy messagebox' commit appears to fail pytest (at least individually).

sumanthvrao commented on tests/ui/test_utils.py:
```diff
@@ -158,17 +162,55 @@ def test_is_muted(mocker, msg, narrow, muted_streams, muted_topics, muted):
+        [],
+        None,
+        False,
+        'Denmark',
```
Sure! I like that we want our code to be more `zulip-terminalish` :D

sumanthvrao commented on tests/ui/test_utils.py:
```diff
@@ -158,17 +162,55 @@ def test_is_muted(mocker, msg, narrow, muted_streams, muted_topics, muted):
+            'recipient_email': 'hamlet@zulip.com',
+            'sender_id': 8
+        },
+        1,
```
Sure. I'll check this out

sumanthvrao commented on zulipterminal/ui_tools/utils.py:
```diff
@@ -20,6 +23,44 @@ def create_msg_box_list(model: Any, messages: Union[None, Iterable[Any]]=None,
+                'display_recipient': stream_caption,
+                'stream_id': model.stream_id,
+                'subject': model.stream_dict.get(model.stream_id)
+                .get('description'),
```
I am working on this. Passing only steam_caption in one place and pm_details in the other seems weird. Shall i make it stream_caption and pass both 'caption', and 'description' of stream from `core.py` ?

sumanthvrao replied:
It looks more common themed now. Hopefully no confusion.

sumanthvrao reviewed (commented):
Some thoughts. Will update PR soon!

sumanthvrao commented:
> NOTE: The 'enable dummy messagebox' commit appears to fail pytest (at least individually).

Fixed it. A indentation error went unnoticed.

sumanthvrao commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -141,6 +141,18 @@ def __init__(self, message: Dict[str, Any], model: Any,
         else:
             raise RuntimeError("Invalid message type")
 
+        if self.message['type'] == 'private':
```
I tried writing a more detailed message. Is that okay?

neiljp replied:
That looks fine :+1:

sumanthvrao commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -141,6 +141,18 @@ def __init__(self, message: Dict[str, Any], model: Any,
+                    for recipient in self.message['display_recipient']
+                    if recipient['email'] != self.model.user_email
+                ))
+
```
The reason i disagree with this is that, they are not independent commits. some part of Enable messagebox code gets tweaked by Refactor commit. The diff will be confusing.

neiljp commented on tests/ui/test_ui_tools.py:
```diff
@@ -1070,6 +1078,58 @@ def test_main_view_generates_PM_header(self, mocker, message,
+                                   'id': None}],
+            'sender_id': 9,
+            'content': "<p> dummy PM message. </p>",
+            'sender_full_name': 'Welcome Bot',
```
Apologies - looking at the code again, I agree; the issue is that `test_init` is not named particularly well. I assumed it tested `__init__` in general, based on the name, but it does not.

sumanthvrao committed:
utils: Build dummy message using parameters passed to create_msg_box.

We use the passed parameters (recipient details, stream details)
while building a dummy message with 'author' as Welcome Bot, 'topic'
as the stream description.

Tests added.

sumanthvrao committed:
core/utils: Changed function calls to create_msg_box.

We send specific information through the function that
makes accessing parameters more easier when building
dummy message;

Added parameters to create_msg_box function definition.

sumanthvrao committed:
boxes: Create dummy message box for empty narrows.

Dummy messages are displayed in narrows with no previous messages.

The action of HotKeys like 'r', 'c', 'z', 'quote', 'reactions' on dummy
messages are different, as they don't have any defined 'context'
(without a real message).

Tests added.

neiljp reviewed (commented):
It is a lot easier to focus on the aspects related to this specific change with the other changes now extracted (and merged) in the other PR :+1:  I definitely like the 'extend MessageBox & test new features', ' extend create_msg_box_list & test new features' and 'integrate' style you now have - it's clear what's changing in each commit, and how tests are adjusting to fix them.

I've left a range of comments on the first three commits, which should keep the development flowing.

My first inclination is to squash the 3rd/4th commits together, since they both integrate the updated features and the 4th feels a little like a fix to the 3rd, but let's see how the PR develops further first.

sumanthvrao commented on tests/ui/test_utils.py:
```diff
@@ -158,17 +162,64 @@ def test_is_muted(mocker, msg, narrow, muted_streams, muted_topics, muted):
+    'Stream narrow - unmuted',
+    'Stream narrow - muted',
+    'Empty stream narrow - dummy message test',
+    'Empty pm narrow - dummy message test'
```
Ah. this command is shorter. I was previously using the --verbose to check the id's. Thanks for the note. Will keep it in mind. :)

sumanthvrao commented on tests/ui/test_ui_tools.py:
```diff
@@ -1111,6 +1111,49 @@ def test_main_view_generates_PM_header(self, mocker, message,
+            assert pm_header == 'Boo Boo'
+        else:
+            stream_header = header.split('>')[0].strip()
+            assert stream_header == 'Foo Foo'
```
Done. :+1:

sumanthvrao commented on tests/ui/test_ui_tools.py:
```diff
@@ -1111,6 +1111,49 @@ def test_main_view_generates_PM_header(self, mocker, message,
+            'sender_id': 9,
+        })
+    ])
+    def test_dummy_message(self, mocker, message):
```
That's pretty long :) Done :+1:

sumanthvrao commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -117,13 +117,14 @@ def keypress(self, size: Tuple[int, int], key: str) -> str:
 class MessageBox(urwid.Pile):
     def __init__(self, message: Dict[str, Any], model: Any,
-                 last_message: Any) -> None:
+                 last_message: Any, is_empty_narrow: bool=False) -> None:
```
Right.

sumanthvrao commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -480,19 +483,22 @@ def keypress(self, size: Tuple[int, int], key: str) -> str:
             self.model.controller.view.middle_column.set_focus('footer')
-        elif is_command_key('QUOTE_REPLY', key):
+        elif is_command_key('QUOTE_REPLY', key) and \
+                not self.is_in_empty_narrow:
```
So a map to what should and should-not be allowed:

Hot keys for dummy_message on PM: 
* r - Y
* R - N (sender is WelcomeBot and not recipient)
* @ - N
* > - N
* c - Y(Same effect as r)
* x - Y
* thumbs up - N
* star status - N
* z/P - Y (works normally)
------
Hot keys for dummy_message on Stream:
* r - Should work the same as c - with no topic filled.
      (To be done).

sumanthvrao replied:
This was discussed on `czo`.

sumanthvrao replied:
@neiljp The thumb-up and star reactions are automatically taken care. It is because the dummy-message is not really in Focus. This is due to addition d3248d7602a9a9c0062e9402b42575f5f0aac622, which checks if `self.focus is not None`.

Perhaps this can be taken care of in a better way. I am hesitant to add more code here as any changes need be done to be done to `views.py` (requiring significant additions atleast).

sumanthvrao commented on zulipterminal/ui_tools/utils.py:
```diff
@@ -1,13 +1,16 @@
-                        last_message: Union[None, Any]=None) -> List[Any]:
+                        last_message: Union[None, Any]=None,
+                        stream_details: Any=None,
+                        pm_details: Any=None) -> List[Any]:
```
Changed the commit message :)

sumanthvrao commented on zulipterminal/ui_tools/utils.py:
```diff
@@ -20,6 +23,43 @@ def create_msg_box_list(model: Any, messages: Union[None, Iterable[Any]]=None,
     last_msg = last_message
     muted_msgs = 0  # No of messages that are muted.
+    if last_msg is not None and last_msg['id'] is None:
+        last_message = None
```
I have forgotten why this was added. It seems to work fine when removed also. Will do some more testing and check it.

sumanthvrao replied:
Ah. So these were added so that 'UP' button for dummy messages would not cause a crash. But it is not necessary now.

sumanthvrao commented on zulipterminal/ui_tools/utils.py:
```diff
@@ -20,6 +23,43 @@ def create_msg_box_list(model: Any, messages: Union[None, Iterable[Any]]=None,
+    # We create a dummy message to show in narrows with
+    # no previous messages, whose author is welcome bot.
+    if (message_list == [] and (stream_details is not None or
+                                pm_details is not None)):
```
I will try adding it. Definitely sounds like good practice :)

sumanthvrao commented on zulipterminal/ui_tools/utils.py:
```diff
@@ -20,6 +23,43 @@ def create_msg_box_list(model: Any, messages: Union[None, Iterable[Any]]=None,
+    # no previous messages, whose author is welcome bot.
+    if (message_list == [] and (stream_details is not None or
+                                pm_details is not None)):
+        msg = {}  # type: Dict[str, Any]
```
Oh yes. This is a mutually exclusive branch now. Initially it was not like this and would have thrown `not defined` error.

sumanthvrao commented on zulipterminal/ui_tools/utils.py:
```diff
@@ -38,11 +78,13 @@ def create_msg_box_list(model: Any, messages: Union[None, Iterable[Any]]=None,
-        last_msg = msg
+        # We don't want to set dummy message as last message.
+        if not is_empty_narrow:
+            last_msg = msg
```
Again I forgot why I added this :'( . Perhaps It was so the append_message event could replace it( At least in one of the cases)

sumanthvrao commented on zulipterminal/core.py:
```diff
@@ -201,9 +213,11 @@ def narrow_to_user(self, button: Any) -> None:
         else:
-            w_list = create_msg_box_list(self.model, msg_id_list)
+            w_list = create_msg_box_list(
+                self.model, msg_id_list, pm_details=pm_details)
```
Yes. However note that `msg_id_list` is None even for `all_messages`. Any refactor suggestion?

sumanthvrao replied:
I changed the code in 2nd commit a little bit. Now it is more understandable. :)

neiljp reviewed (commented):
Thanks for sticking with this :)

I think the `msg` change lead to a regression, eg. if you press down at the 'end' of the stream. That may just be caused by removing the extra condition from the previous version. That is, calling `create_msg_box_list` with no messages but it treats that as assuming `pm_details` or `stream_details`, which may not be the case there?

sumanthvrao commented on tests/ui/test_ui_tools.py:
```diff
@@ -1111,6 +1111,52 @@ def test_main_view_generates_PM_header(self, mocker, message,
+            'stream_id': 8,
+            'index_value': 0,
+            'split_character': '>',
+            'assert_string': 'Foo Foo'
```
Done :+1: . This was quite useful :)

sumanthvrao commented on zulipterminal/ui_tools/utils.py:
```diff
@@ -20,6 +23,44 @@ def create_msg_box_list(model: Any, messages: Union[None, Iterable[Any]]=None,
+        if stream_details is not None:
+            assert len(stream_details) == 2
+        if pm_details is not None:
+            assert len(pm_details) == 3
```
Got it. Apologies I misunderstood you the previous time. I made use of `messages` itself in the assert statement. However I had to reorder the commits and add couple of changes to `views.py`. I have made this a separate commit so no worries. :)

sumanthvrao commented on zulipterminal/ui_tools/utils.py:
```diff
@@ -20,6 +23,44 @@ def create_msg_box_list(model: Any, messages: Union[None, Iterable[Any]]=None,
+                'stream_id': model.stream_id,
+                'subject': stream_details['description']
+                }
+        elif pm_details is not None:
```
I added the assert statement you suggested earlier. So we use the `else` as you suggest

sumanthvrao committed:
views: Handles up and down keys on dummy messages.

Handles 'Up' and 'Down' keys on dummy message; Otherwise
it would raise an exception as there are neither old/new
messages above/below the dummy message.

sumanthvrao commented on tests/model/test_model.py:
```diff
@@ -538,6 +538,51 @@ def test_append_message_event_flags(self, mocker, model):
         set_count.assert_called_once_with([event['message']['id']],
                                           self.controller, 1)
 
+    @pytest.mark.parametrize('message', [
```
Right. makes more sense :)

sumanthvrao commented on tests/model/test_model.py:
```diff
@@ -538,6 +538,51 @@ def test_append_message_event_flags(self, mocker, model):
+            'type': 'private',
+            'id': 0,
+            'display_recipient': [{'email': 'boo@zulip.com'}]
+        }, 1)
```
a) oops. That was by mistake.
b) Fundamentally we want to show that either the dummy message / the sent message (replacing the dummy message) are in the narrow. Not both of them (assert == 2) or none of them (asser == 1).
c) I understood why you are asking to assert narrows. This is related when the narrows seemed to switch when you were initially testing the PR (after i sent a PM). Well in that case, actually, the dummy message got deleted. It looked like the narrow changed.

The login in `append_message` now handles these cases.
I have removed the extra test and moved the 1 inside the test.

sumanthvrao commented:
@neiljp FYI.
We had misunderstood the fact that `create_msg_box` was called when 
 ```
message_list != [] or not model.narrow or(
    (stream_details is not None and pm_details is None) or
    (pm_details is not None and stream_details is None))
```
was the true. Hence the above assertion made sense initially. But upon testing further I got quite a few regressions with this assert statement. 
Eg:
`w_list = create_msg_box_list(self.model, msg_id_list)` 
of `narrow_to_topic` in core.py and 
``` 
message_list = create_msg_box_list(self.model, new_ids, last_message=last_message)
```
of `load_new_message` in `views.py`... and some more call `create_msg_box` with `message_list = []`. This was the reason I decided to remove the assert statement.

sumanthvrao committed:
helper: Prevent narrows from being marked 'M'.

We mark new PMs which replace the dummy message
as being 'unread'. This avoids marking 'All messages',
'All Private' and other PM narrows with 'M'.

neiljp commented on zulipterminal/model.py:
```diff
@@ -473,6 +473,36 @@ def append_message(self, event: Event) -> None:
+                        if possible_dummy_message.get('display_recipient') == \
+                                response['display_recipient']:
+                            dummy_message_found = False
+                            del self.msg_list.log[0]
```
This is probably the code that feels the most like it isn't quite there yet. The tests don't cover multiple recipients, which might make this clearer and cover more cases? Also, if you could integrate it with the cases below, which should cover existing cases, then that may make it easier? I'm not sure if you tried that first?

sumanthvrao replied:
Yup. I have integrated it with the cases below. Further, few other PRs submitted (#362, #364 and #365) have helped simplify the complexity of this commit to a great degree.

neiljp reviewed (commented):
@sumanthvrao Returning to this PR I remember how complex & involved it is!

I've done some manual testing, which uncovered that `s` or `S` from a private dummy message goes to all messages, which seems unexpected!

I'm a little unsure of the (new?) last commit (Prevent narrows from being marked 'M'); it looks like addressing a symptom (forcing the count up), rather than the problem (not increasing it in the first place). I'm not precisely sure what triggers this case (the commit message isn't entirely obvious) - is this `set_count` being called by `Model.set_message_ids_as_read`? If so, can we just fix the problem at source, which is probably in `MessageView.read_message`?

The commit flow looks pretty good, though I'm marginally wary of the 2nd commit being present before the parameters are added to `create_msg_box_list`, though I think we can't actually trigger that in the code at that point, perhaps? Either way, it seems strange adding an invalid call to a function before the function is updated to allow the extra parameters. Do you see my concern?

sumanthvrao commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Dummy messages in empty narrows](https://trello.com/c/pBjuiOCw/5-dummy-messages-in-empty-narrows)

sumanthvrao commented on tests/ui/test_ui_tools.py:
```diff
@@ -1114,6 +1114,49 @@ def test_main_view_generates_PM_header(self, mocker, message,
+        header = view_components[0].original_widget.get_text()[0].strip()
+        header_message = (header.split(split_character)
+                          [index_value].strip())
+        assert header_message == assert_string
```
Yup. Done. Thanks for spotting this :)

sumanthvrao commented on zulipterminal/ui_tools/utils.py:
```diff
@@ -7,7 +7,9 @@
-                        last_message: Union[None, Any]=None) -> List[Any]:
+                        last_message: Union[None, Any]=None,
+                        stream_details: Any=None,
+                        pm_details: Any=None) -> List[Any]:
```
Yup :+1: . Done.

sumanthvrao commented on zulipterminal/ui_tools/views.py:
```diff
@@ -87,6 +90,9 @@ def load_new_messages(self, anchor: int) -> None:
 
+        # This for the dummy message
+        if current_ids == set():
+            return
```
I did try commenting it and it did not produce an exceptions. On the other hand, I seem to have forgotten why this was added in the first place. To be on the safe side, I followed the third approach here, mainly to rewrite the code similar to the first part - check for log; check for top message in the log; and check if it is a dummy message. Thanks for this :+1:

sumanthvrao commented on tests/ui/test_utils.py:
```diff
@@ -144,5 +186,7 @@ def test_create_msg_box_list(mocker, narrow, messages, focus_msg_id,
+    return_value = create_msg_box_list(model, messages, focus_msg_id,
+                                       stream_details=stream_details,
+                                       pm_details=pm_details)
     assert len(return_value) == len_w_list
```
I did add another assert here to check the number of times `MessageBox` is called.  Further, tests have been added in previous commits specifically in `test_main_view_generates_dummy_message` which tests for assumptions in empty narrow.  Further asserts to check assumptions about the narrows would complicate this test to a high degree.

sumanthvrao commented on tests/model/test_model.py:
```diff
@@ -541,6 +541,56 @@ def test_append_message_event_flags(self, mocker, model):
+            'id': 0,
+            'display_recipient': [{'email': 'bar@zulip.com'}]
+        })
+    ])
```
Done. :+1:

sumanthvrao committed:
model/views: Remove dummy message during append_message event.

This commit handles the logic of replacing the dummy
message when an incoming message, meant for the same
narrow appears.

* Ensure dummy message for stream and user PM is replaced
   when the first message in those respective narrows appear.
* Ensure group PM messages do not replace dummy message in user PM narrow.
* Ensure dummy message for self-message narrow is replaced.

Tests added.

sumanthvrao commented:
@neiljp Thanks for patiently reviewing this again. I apologise for the delay. Here is a small CHANGELOG from the previous iteration:
*  's' and 'S' for dummy messages have been handled. 
*  Extra assert has been added for commit #1 tests.
*  Extra assert to commit #4, to test no of times MessageBox is called.
*  Re-worded the code surrounding `load_new_messages` for dummy messages.
*  Added Ids for test - `test_append_message_with_dummy_previous_message`
*  The last commit is a fix for a bug which can be reproduced by sending a message:
1. Login as Iago from web-app.
2. Login as Cordelia from ZT. 
3. Ensure there is no convo between Cordelia and Iago. Dummy message should be shown
4. Send a message from webapp to Cordelia.
Observe 'M' being marked on 'All Messages', 'All Private Messages', 'Iago user'.

*Need clarification on* :
Change requested for commit #2 
```
The commit flow looks pretty good, though I'm marginally wary of the 2nd commit  ...
```

sumanthvrao commented:
@neiljp I rebased this and resolved the conflicts. Its now good for a review. :)

sumanthvrao committed:
boxes/buttons: Add user_name field to UserButton.

This is necessary in boxes & buttons which act with the
associated narrowing methods in core.py

This addition is made use of in successive commit.
